docs/design_notes/lift.md

```diff
@@ -0,0 +1,338 @@
+A function output can only change when the input arguments change.
+Therefore, a function is garantueed to behave deterministicly.
+
+But pure functions offer another big advantage to JAX specifically: They enable functional transformations.
+For example `jax.vmap(f)` will vectorize a function `f`.
+Because `f` cannot have side-effects the vectorized/parallel version of `f` is well-defined. To see why we need this restriction, consider what happens if `f` would increment a counter or draw a random number.
+Would `f` draw the same or a different random number for each item in the vector?
+Would each item in the batch have its own counter or is the counter shared among the items?
+And in what order is the counter incremented if `f` is computed in parallel?
+The anser to all these questions is "it depends".
+The behavior is ambigious and the functional constraint elegantly avoids this problem.
+
+Flax introduces a safe way to have limited randomness and stateful variables in a JAX compatible form.
+The reason state in Flax is not problematic is because it is local. Inside a flax `Module` there are variables and PRNG sequences,
+but on the outside there are only JAX Arrays and PRNG keys.
+
+For most use cases Flax be used to define models in a stateful way.
+Becuase a `Module` behaves pure functional externally  we can utilize the full power of JAX with all of its transformations.
+There are however cases where we want to have the best of both worlds by using transformations and Module together.
+This design note explains how we extend JAX functional transformation to work on modules that have internal state and randomness.
+
+
+## Functionalization
+
+Before we jump into the details let's consider a simple example where we would like to use `vmap` inside a `Module`.
+
+First we define a simple MLP without any transformations:
+```
+import jax
+from jax import random, numpy as jnp
+from flax import linen as nn
+
+class MLP(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    h = nn.Dense(4, name='hidden')(xs)
+    h = nn.relu(h)
+    return nn.Dense(1, name='out')(h)
+```
+
+Now what if I want to have seperate MLP parameters for each item in `xs`?
+If this where "vanilla JAX" we could imagene writing something like `jax.vmap(apply_mlp)(mlp_params, xs)`.
+But doing something like this in linen will actually fail:
+```
+class NaiveVmapMLP(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    mlps = MLP()
+    return jax.vmap(lambda mlp, x: mlp(x))(mlps, xs)  # fails
+```
+
+JAX will refuse to transform the `mlps` because its not a JAX Array or a simple container of arrays.
+We can't really blame JAX for refusing this dirty, underspecified task.
+After all it's not even clear what should happen here.
+The parameters inside MLP are not even initialized yet and we will need a seperate PRNG key for each group of parameters.
+
+We can fix this problem by first turning `MLP` into a pure init and apply function.
```

Comment:
Suggestion (similar to how you used the formatting below):

```suggestion
We can fix this problem by first turning `MLP` into a pure `init` and `apply` function.
```

(or `.init`, `.apply`)

Reply:
Here we refer to an init/apply function not a particular one or the method that creates them.  So I think this should be text not code

Reply:
Understood, thanks for clarifying this @jheek